# **Pandas Case Study - Indian Startup Funding**

In [1]:
import pandas as pd

In [2]:
# load startup_funding.csv
df = pd.read_csv('startup_funding.csv')
df.head()

Sr No Date dd/mm/yyyy                  Startup Name    Industry Vertical  \
0      1      09/01/2020                        BYJU’S               E-Tech   
1      2      13/01/2020                        Shuttl       Transportation   
2      3      09/01/2020                     Mamaearth           E-commerce   
3      4      02/01/2020  https://www.wealthbucket.in/              FinTech   
4      5      02/01/2020                        Fashor  Fashion and Apparel   

                             SubVertical City  Location  \
0                             E-learning      Bengaluru   
1              App based shuttle service        Gurgaon   
2  Retailer of baby and toddler products      Bengaluru   
3                      Online Investment      New Delhi   
4            Embroiled Clothes For Women         Mumbai   

              Investors Name       InvestmentnType Amount in USD Remarks  
0    Tiger Global Management  Private Equity Round  20,00,00,000     NaN  
1  Susquehanna Growth Equity              Series C     80,48,394     NaN  
2      Sequoia Capital India              Series B   1,83,58,860     NaN  
3             Vinod Khatumal          Pre-series A     30,00,000     NaN  
4    Sprout Venture Partners            Seed Round     18,00,000     NaN

In [3]:
df.sample(5)

Sr No Date dd/mm/yyyy       Startup Name  Industry Vertical  \
363     364      09/02/2018          WebEngage         Technology   
821     822      15/05/2017              Zapty         Technology   
1556   1557      20/07/2016            Tax2Win  Consumer Internet   
872     873      07/04/2017  Twenty Two Motors         Technology   
1963   1964      17/02/2016         Timemytask  Consumer Internet   

                                  SubVertical City  Location  \
363   Marketing Cloud for consumer businesses         Mumbai   
821                   Project Management tool      Bangalore   
1556                 Online Tax Filing Portal         Jaipur   
872                Smart Scooter manufacturer        Gurgaon   
1963         On-Demand home cleaning services         Mumbai   

                    Investors Name InvestmentnType Amount in USD Remarks  
363                 Social Capital  Private Equity           NaN     NaN  
821             IdeaSpring Capital    Seed Funding      5,00,000     NaN  
1556         Undisclosed Investors  Private Equity     10,00,000     NaN  
872   Ishwar Singh,Farhaan Shabbir    Seed Funding           NaN     NaN  
1963                   Lead Angels    Seed Funding           NaN     NaN

In [4]:
df.shape

(3044, 10)

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3044 entries, 0 to 3043
Data columns (total 10 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   Sr No              3044 non-null   int64 
 1   Date dd/mm/yyyy    3044 non-null   object
 2   Startup Name       3044 non-null   object
 3   Industry Vertical  2873 non-null   object
 4   SubVertical        2108 non-null   object
 5   City  Location     2864 non-null   object
 6   Investors Name     3020 non-null   object
 7   InvestmentnType    3040 non-null   object
 8   Amount in USD      2084 non-null   object
 9   Remarks            419 non-null    object
dtypes: int64(1), object(9)
memory usage: 237.9+ KB


In [6]:
# We can observe here there are many "NaN" values and 
# the data provided is not very clean to start the analysis

## ``Preprocessing``

#### 1. Make `Sr No` as index

In [7]:
df.drop(columns=['Remarks'], inplace=True)

In [8]:
df.head(3)

Sr No Date dd/mm/yyyy Startup Name Industry Vertical  \
0      1      09/01/2020       BYJU’S            E-Tech   
1      2      13/01/2020       Shuttl    Transportation   
2      3      09/01/2020    Mamaearth        E-commerce   

                             SubVertical City  Location  \
0                             E-learning      Bengaluru   
1              App based shuttle service        Gurgaon   
2  Retailer of baby and toddler products      Bengaluru   

              Investors Name       InvestmentnType Amount in USD  
0    Tiger Global Management  Private Equity Round  20,00,00,000  
1  Susquehanna Growth Equity              Series C     80,48,394  
2      Sequoia Capital India              Series B   1,83,58,860

#### 2. `Drop remarks col` --> contains only 419 non-null values out of 3044

In [9]:
df.set_index('Sr No', inplace=True)

In [10]:
df.head(2)

Date dd/mm/yyyy Startup Name Industry Vertical  \
Sr No                                                  
1          09/01/2020       BYJU’S            E-Tech   
2          13/01/2020       Shuttl    Transportation   

                     SubVertical City  Location             Investors Name  \
Sr No                                                                        
1                     E-learning      Bengaluru    Tiger Global Management   
2      App based shuttle service        Gurgaon  Susquehanna Growth Equity   

            InvestmentnType Amount in USD  
Sr No                                      
1      Private Equity Round  20,00,00,000  
2                  Series C     80,48,394

* 3. Lets `change the names of cols` for our convinence

In [11]:
df.rename(columns={
    'Date dd/mm/yyyy': 'date',
    'Startup Name': 'startup',
    'Industry Vertical': 'vertical',
    'SubVertical': 'subvertical',
    'City  Location': 'city',
    'Investors Name': 'investors',
    'InvestmentnType': 'round',
    'Amount in USD': 'amount'
}, inplace=True)

In [12]:
df.head(3)

date    startup        vertical  \
Sr No                                          
1      09/01/2020     BYJU’S          E-Tech   
2      13/01/2020     Shuttl  Transportation   
3      09/01/2020  Mamaearth      E-commerce   

                                 subvertical       city  \
Sr No                                                     
1                                 E-learning  Bengaluru   
2                  App based shuttle service    Gurgaon   
3      Retailer of baby and toddler products  Bengaluru   

                       investors                 round        amount  
Sr No                                                                 
1        Tiger Global Management  Private Equity Round  20,00,00,000  
2      Susquehanna Growth Equity              Series C     80,48,394  
3          Sequoia Capital India              Series B   1,83,58,860

#### 4. Convert amount from `UDS to Rs`:
* a. Convert string to float
* b. Convert usd to rs

In [13]:
df['amount'].isnull().sum() # There are 960 null values.

960

In [14]:
# Since there are null values we cannot convert it into float. 
# Therefore, make null to '0', which means there are not disclosed

In [15]:
df['amount'] = df['amount'].fillna('0')

In [16]:
df['amount'].isnull().sum()

0

In [17]:
# convert to float.
# Before that remove ','. Ex: 20,000
df['amount'] =df['amount'].str.replace(',', '')

In [18]:
df['amount'].str.contains(',').sum() # checking if any elements contains ','

0

In [19]:
# check whether there are any alphabets with numeric
df['amount'].str.isalpha().sum()

7

In [20]:
# check how many are numeric
df['amount'].str.isnumeric().sum()

3022

In [24]:
# Finding non numeric values
df[~df['amount'].str.isnumeric()]['amount']

Sr No
21             undisclosed
35                 unknown
59             Undisclosed
77              4889975.54
82              2739034.68
84             15109500.00
90             undisclosed
92             undisclosed
110              14342000+
113            Undisclosed
140            Undisclosed
2603    \\xc2\\xa020000000
2604    \\xc2\\xa016200000
2605         \\xc2\\xa0N/A
2606         \\xc2\\xa0N/A
2607      \\xc2\\xa0600000
2608      \\xc2\\xa0685000
2609    \\xc2\\xa019350000
2610     \\xc2\\xa05000000
2611    \\xc2\\xa010000000
2612         \\xc2\\xa0N/A
2614         \\xc2\\xa0N/A
Name: amount, dtype: object

In [23]:
df['amount'][84]

'15109500.00'

In [25]:
# replace non-numeric values with '0'
df['amount'] = df['amount'].apply(lambda x: '0' if not str(x).isnumeric() else x)

In [26]:
df[~df['amount'].str.isnumeric()]['amount'].sum()

0

In [27]:
# convert to float
df['amount'] = df['amount'].astype('float')

In [28]:
# function to convert usd to rs
def usd_to_rs(amt):
    rs = 82.53 * amt
    # we want to return in crores
    return rs/10000000

In [29]:
df['amount'] = df['amount'].apply(usd_to_rs)

In [30]:
df['amount'][:3]

Sr No
1    1650.600000
2      66.423396
3     151.515672
Name: amount, dtype: float64

In [31]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3044 entries, 1 to 3044
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   date         3044 non-null   object 
 1   startup      3044 non-null   object 
 2   vertical     2873 non-null   object 
 3   subvertical  2108 non-null   object 
 4   city         2864 non-null   object 
 5   investors    3020 non-null   object 
 6   round        3040 non-null   object 
 7   amount       3044 non-null   float64
dtypes: float64(1), object(7)
memory usage: 278.6+ KB


* 5. Convert `date col` to date time obj 

In [32]:
# Filter out elements with format '%m/%d/%Y' and obtain non-matching elements
pattern = r'\b\d{2}/\d{2}/\d{4}\b' # regex pattern
filtered_df = df[~df['date'].str.match(pattern, na=False)]['date']

print(filtered_df)
print(filtered_df.size)

Sr No
193               05/072018
2572              01/07/015
2607    \\xc2\\xa010/7/2015
2776             12/05.2015
2777             12/05.2015
2832             13/04.2015
3012             15/01.2015
3030            22/01//2015
Name: date, dtype: object
8


In [ ]:
# # # Preprocess the date column
# df['date'] = df['date'].str.replace(r'[^0-9/]', '')  # Remove unwanted characters

# # Extract day, month, and year components
# day = df['date'].str.extract(r'(\d{1,2})/')  # Extract day component
# month = df['date'].str.extract(r'/(?P<month>\d{1,2})/')  # Extract month component
# year = df['date'].str.extract(r'/(?P<year>\d{4})$')  # Extract year component

# # Combine day, month, and year into a consistent format
# # ...............

# # Filter out valid dates
# preprocessed_df = df[df['date'].notnull()]['date']

# print(preprocessed_df)

In [33]:
# Since there are only 8 different formated values, lets use 'coerce'
# df['date'] =
df['date'] = pd.to_datetime(df['date'], errors='coerce')

In [34]:
df.head(3)

date    startup        vertical  \
Sr No                                         
1     2020-09-01     BYJU’S          E-Tech   
2     2020-01-13     Shuttl  Transportation   
3     2020-09-01  Mamaearth      E-commerce   

                                 subvertical       city  \
Sr No                                                     
1                                 E-learning  Bengaluru   
2                  App based shuttle service    Gurgaon   
3      Retailer of baby and toddler products  Bengaluru   

                       investors                 round       amount  
Sr No                                                                
1        Tiger Global Management  Private Equity Round  1650.600000  
2      Susquehanna Growth Equity              Series C    66.423396  
3          Sequoia Capital India              Series B   151.515672

* 6. Drop rows with missing values

In [35]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3044 entries, 1 to 3044
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   date         3040 non-null   datetime64[ns]
 1   startup      3044 non-null   object        
 2   vertical     2873 non-null   object        
 3   subvertical  2108 non-null   object        
 4   city         2864 non-null   object        
 5   investors    3020 non-null   object        
 6   round        3040 non-null   object        
 7   amount       3044 non-null   float64       
dtypes: datetime64[ns](1), float64(1), object(6)
memory usage: 278.6+ KB


In [36]:
# Except subvertical drop rows from every other cols with missing values
df = df.dropna(subset=['date', 'startup', 'vertical', 'city', 'investors', 'round', 'amount'])

In [37]:
df.head()

date                       startup             vertical  \
Sr No                                                                 
1     2020-09-01                        BYJU’S               E-Tech   
2     2020-01-13                        Shuttl       Transportation   
3     2020-09-01                     Mamaearth           E-commerce   
4     2020-02-01  https://www.wealthbucket.in/              FinTech   
5     2020-02-01                        Fashor  Fashion and Apparel   

                                 subvertical       city  \
Sr No                                                     
1                                 E-learning  Bengaluru   
2                  App based shuttle service    Gurgaon   
3      Retailer of baby and toddler products  Bengaluru   
4                          Online Investment  New Delhi   
5                Embroiled Clothes For Women     Mumbai   

                       investors                 round       amount  
Sr No                                                                
1        Tiger Global Management  Private Equity Round  1650.600000  
2      Susquehanna Growth Equity              Series C    66.423396  
3          Sequoia Capital India              Series B   151.515672  
4                 Vinod Khatumal          Pre-series A    24.759000  
5        Sprout Venture Partners            Seed Round    14.855400

In [38]:
df.shape

(2833, 8)

In [39]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2833 entries, 1 to 2873
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   date         2833 non-null   datetime64[ns]
 1   startup      2833 non-null   object        
 2   vertical     2833 non-null   object        
 3   subvertical  2079 non-null   object        
 4   city         2833 non-null   object        
 5   investors    2833 non-null   object        
 6   round        2833 non-null   object        
 7   amount       2833 non-null   float64       
dtypes: datetime64[ns](1), float64(1), object(6)
memory usage: 199.2+ KB
